### PR TITLE
Add docs for kNN search endpoint

### PR DESCRIPTION
**Pull request by jtibshirani**
This commit adds docs for the new `_knn_search` endpoint.

It focuses on being an API reference and is light on details in terms of how
exactly the kNN search works, and how the endpoint contrasts with
`script_score` queries. We plan to add a high-level guide on kNN search that
will explain this in depth.

Relates to #78473.

### CODE REVIEW

**Comment by jtibshirani** [comment]
This is a "rough draft" and isn't ready for a full review. I opened it to get feedback on one point: there is a lot of overlap with the `_search` endpoint in terms of request and response sections. I started to go down the path of tagging several sections in `search.asciidoc` so I could include them in `knn-search.asciidoc` with no copying. Does this seem like a good direction? There are still a few left to tag (`took`, `_shards`, `hits`...)

File: docs/reference/search/search.asciidoc [hunk 1]
```diff
@@ -92,7 +92,7 @@ include::{es-repo-dir}/rest-api/common-parms.asciidoc[tag=df]
 
 `docvalue_fields`::
 (Optional, string) A comma-separated list of fields to return as the docvalue
-representation of a field for each hit.
+representation of a field for each hit. See <<docvalue-fields>>.
```
Comment (jtibshirani) [inline docs/reference/search/search.asciidoc:95]: This isn't closely related to the change, but I saw the opportunity to refactor the field retrieval docs a bit.

File: docs/reference/mapping/types/dense-vector.asciidoc [hunk 2]
```diff
@@ -8,16 +8,19 @@
 
 The `dense_vector` field type stores dense vectors of float values.
 
-You can use `dense_vector` fields in
-<<query-dsl-script-score-query,`script_score`>> queries to score documents.
-They can also be indexed to support efficient k-nearest neighbor search. Dense
-vector fields do not support aggregations, sorting, or other query types.
+Dense vector fields can be used in the following ways:
+* In <<query-dsl-script-score-query,`script_score`>> queries to score
```
Comment (jrodewig) [inline docs/reference/mapping/types/dense-vector.asciidoc:12]: I think asciidoc requires an empty line to render this correctly.
```suggestion

* In <<query-dsl-script-score-query,`script_score`>> queries to score
```

File: docs/reference/search/knn-search.asciidoc [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+
+[role="child_attributes"]
+[[knn-search-api-response-body]]
+==== {api-response-body-title}
+
+A kNN search response has the exact same structure as a
+<<search-api-response-body, search response>>. The
+<<knn-search-api-response-body-score,`_score`>> of each hit is determined by
+the similarity between the query and document vectors. See
+<<dense-vector-similarity, `similarity`>> for details.
```
Comment (jrodewig) [inline docs/reference/search/knn-search.asciidoc:124]: I think this short paragraph is sufficient. We do something similar in the [EQL async search API docs](https://www.elastic.co/guide/en/elasticsearch/reference/master/get-async-eql-search-api.html#get-async-eql-search-api-response-body). I wouldn't go through the hassle of tagging and including each field.

We may want to include a short bulleted list detailing any differences. For example, `hits` means something different for kNN.

**Review by jrodewig — COMMENTED** [review]
This looks good so far! I'd take the same approach you have.

Tag + reuse is probably a good fit for the query + request body params, but I think it's overkill for the response. We should just link users to the existing search API response body docs and concisely point out any differences.

File: docs/reference/search/knn-search.asciidoc (hunk 3 above)
Comment (jtibshirani) [reply docs/reference/search/knn-search.asciidoc:124]: This is a good idea, I'll just add a section highlighting the differences.

File: docs/reference/search/knn-search.asciidoc [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+or alias.
+
+[[knn-search-api-desc]]
+==== {api-description-title}
+
+The kNN search API performs a k-nearest neighbor search on a
+<<dense-vector,`dense_vector`>> field. Given a query vector, it finds the _k_
+closest vectors and returns those documents as search hits. To be used in the
+kNN search API, the `dense_vector` field must
+<<index-vectors-knn-search, enable indexing>>.
```
Comment (jrodewig) [inline docs/reference/search/knn-search.asciidoc:61]: Not a huge deal, but I'd consider moving the last sentence into the `knn.field` param definition. It's sort of a requirement for that param.

I overlooked this when we created the dense vector docs, but it may be good to note there that the `index` parameter isn't dynamic/updateable.

File: docs/reference/search/knn-search.asciidoc [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+[[knn-search-api-path-params]]
+==== {api-path-parms-title}
+
+`<target>`::
+(Optional, string) Comma-separated list of data streams, indices, and aliases
+to search. Supports wildcards (`*`). To search all data streams and indices,
+use `*` or `_all`.
+
+NOTE: The kNN search API does not support remote indices (as in cross-cluster
+search).
```
Comment (jrodewig) [inline docs/reference/search/knn-search.asciidoc:77]:
```suggestion
use `*` or `_all`.
+
NOTE: The kNN search API doesn't support {ccs}.
```

File: docs/reference/search/knn-search.asciidoc [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+==== {api-request-body-title}
+
+`knn`::
+(Required, object) Defines the kNN query to run.
++
+.Properties of `knn` object
+[%collapsible%open]
+====
+`field`::
+(Required, string) The name of the vector field to search against.
```
Comment (jrodewig) [inline docs/reference/search/knn-search.asciidoc:96]: See above comment about noting the `dense_vector` and indexing requirement here.

File: docs/reference/search/knn-search.asciidoc [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+`query_vector`::
+(Required, array of floats) The query vector.
```
Comment (jrodewig) [inline docs/reference/search/knn-search.asciidoc:99]: Should we note this must be a `dense_vector` value?
```suggestion
(Required, array of floats) Query vector.
```

File: docs/reference/search/knn-search.asciidoc [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+`k`::
+(Required, integer) The final number of nearest neighbors to return as top
+hits. This value must be less than `num_candidates`.
```
Comment (jrodewig) [inline docs/reference/search/knn-search.asciidoc:103]:
```suggestion
(Required, integer) Number of nearest neighbors to return as top hits. Must be
less than `num_candidates`.
```

File: docs/reference/search/knn-search.asciidoc [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+`num_candidates`::
+(Required, integer) The number of nearest neighbor candidates to consider per
+shard. Increasing `num_candidates` tends to improve the accuracy of the final
+`k` results. This value cannot exceed 10,000.
```
Comment (jrodewig) [inline docs/reference/search/knn-search.asciidoc:108]: We don't really describe the high-level process for gathering candidates from each shard and then collating the results. IIUC that and HSNW are the two main factors that can affect accuracy.

Should we save that for the high-level guide or mention in the description?

```suggestion
(Required, integer) Number of nearest neighbor candidates to consider per shard.
Cannot exceed `10000`. A higher value improves the accuracy of the results at
the cost of a slower search speed.
```

File: docs/reference/search/knn-search.asciidoc [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+include::{es-repo-dir}/search/search.asciidoc[tag=stored-fields-def]
+include::{es-repo-dir}/search/search.asciidoc[tag=source-filtering-def]
+
+[role="child_attributes"]
+[[knn-search-api-response-body]]
+==== {api-response-body-title}
+
+A kNN search response has the exact same structure as a
+<<search-api-response-body, search response>>. However, certain sections take
+on a meaning specific to kNN search:
```
Comment (jrodewig) [inline docs/reference/search/knn-search.asciidoc:122]:
```suggestion
<<search-api-response-body, search response>>. However, certain sections have
a meaning specific to kNN search:
```

File: docs/reference/search/knn-search.asciidoc [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+* The <<search-api-response-body-score,document `_score`>> is determined by
+the similarity between the query and document vector. See
+<<dense-vector-similarity, `similarity`>> for details.
```
Comment (jrodewig) [inline docs/reference/search/knn-search.asciidoc:126]:
```suggestion
<<dense-vector-similarity, `similarity`>>.
```

File: docs/reference/search/knn-search.asciidoc [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+
+[role="child_attributes"]
+[[knn-search-api-response-body]]
+==== {api-response-body-title}
+
+A kNN search response has the exact same structure as a
+<<search-api-response-body, search response>>. However, certain sections take
+on a meaning specific to kNN search:
+
+* The <<search-api-response-body-score,document `_score`>> is determined by
```
Comment (jrodewig) [inline docs/reference/search/knn-search.asciidoc:124]: Tried to reword this to avoid passive voice. Not a big deal if left as is tho.
```suggestion
* The <<search-api-response-body-score,document `_score`>> measures
```

File: docs/reference/search/knn-search.asciidoc [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+the similarity between the query and document vector. See
+<<dense-vector-similarity, `similarity`>> for details.
+* The `hits.total` object contains the total number of nearest neighbor
+candidates considered, which is `num_candidates * num_shards`. The `relation`
```
Comment (jrodewig) [inline docs/reference/search/knn-search.asciidoc:128]:
```suggestion
candidates considered, which is `num_candidates * num_shards`. `hits.total.relation`
```

File: docs/reference/search/search.asciidoc (hunk 1 above)
Comment (jrodewig) [reply docs/reference/search/search.asciidoc:95]: These look great. I particularly like use "field pattern."

File: docs/reference/search/search.asciidoc [hunk 14]
```diff
@@ -425,8 +430,21 @@ returns the tile as a base64-encoded string.
...
 ========
 --
-// end::fields-api-props[]
+// end::fields-param-props[]
 ====
+// end::fields-param-def[]
+
+[[search-stored-fields-param]]
+// tag::stored-fields-def[]
+`stored_fields`::
```
Comment (jrodewig) [inline docs/reference/search/search.asciidoc:439]: Oof. I can't believe this wasn't documented. Thanks for catching it.

File: docs/reference/search/knn-search.asciidoc [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+include::{es-repo-dir}/search/search.asciidoc[tag=docvalue-fields-def]
+include::{es-repo-dir}/search/search.asciidoc[tag=stored-fields-def]
+include::{es-repo-dir}/search/search.asciidoc[tag=source-filtering-def]
+
+[role="child_attributes"]
+[[knn-search-api-response-body]]
+==== {api-response-body-title}
+
+A kNN search response has the exact same structure as a
+<<search-api-response-body, search response>>. However, certain sections take
```
Comment (jrodewig) [inline docs/reference/search/knn-search.asciidoc:121]: I'd specifically mention search API here as we have a number of search-related APIs now.
```suggestion
A kNN search response has the same structure as a
<<search-api-response-body, search API response>>. However, certain sections take
```

File: docs/reference/search/knn-search.asciidoc [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+`num_candidates`::
+(Required, integer) The number of nearest neighbor candidates to consider per
+shard. Increasing `num_candidates` tends to improve the accuracy of the final
+`k` results. This value cannot exceed 10,000.
+====
+
+include::{es-repo-dir}/search/search.asciidoc[tag=fields-param-def]
+include::{es-repo-dir}/search/search.asciidoc[tag=docvalue-fields-def]
+include::{es-repo-dir}/search/search.asciidoc[tag=stored-fields-def]
+include::{es-repo-dir}/search/search.asciidoc[tag=source-filtering-def]
```
Comment (jrodewig) [inline docs/reference/search/knn-search.asciidoc:114]: We typically try to order parameters alphabetically, but I kinda like this sort order.

**Review by jrodewig — APPROVED** [review]
This looks great. I left some comments and suggestions, but feel free to disregard as wanted. Nothing is blocking, aside from a syntax error in the `stored_fields` def. Thanks!

File: docs/reference/search/knn-search.asciidoc (hunk 7 above)
Comment (mayya-sharipova) [reply docs/reference/search/knn-search.asciidoc:99]: May be we can add that `query_vector` must have the same dims as an indexed field it searches against. Although it is also kind of obvious, so not sure if it is worth to add.

File: docs/reference/search/knn-search.asciidoc (hunk 9 above)
Comment (mayya-sharipova) [reply docs/reference/search/knn-search.asciidoc:108]: I agree with @jrodewig it would be nice to add more details here how this `num_candidates` works, for example: "{es}  collects from each shard  the top `num_candidates` results, and then merges collected from the shards results to get the top k results.  Increasing `num_candidates` tends to improve the accuracy of the final `k` results..."

**Review by mayya-sharipova — APPROVED** [review]
@jtibshirani Thanks, a great addition! I've left a couple of small comments.

File: docs/reference/search/knn-search.asciidoc (hunk 5 above)
Comment (jtibshirani) [reply docs/reference/search/knn-search.asciidoc:77]: This is much simpler! I hadn't realized we used "CCS" in the context of APIs besides `_search` (field caps, now kNN, etc.)

File: docs/reference/search/knn-search.asciidoc (hunk 15 above)
Comment (jtibshirani) [reply docs/reference/search/knn-search.asciidoc:121]: 👍  makes sense.

File: docs/reference/mapping/types/dense-vector.asciidoc [hunk 17]
```diff
@@ -6,18 +6,21 @@
...
+vector fields can be used in the following ways:
 
-You can use `dense_vector` fields in
-<<query-dsl-script-score-query,`script_score`>> queries to score documents.
-They can also be indexed to support efficient k-nearest neighbor search. Dense
-vector fields do not support aggregations, sorting, or other query types.
+* In <<query-dsl-script-score-query,`script_score`>> queries, to score
+documents matching a filter
+* In the <<knn-search, kNN search API>>, to find the _k_ most similar vectors
+to a query vector
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply docs/reference/mapping/types/dense-vector.asciidoc:15]: I don't usually put periods at the end of bulleted items unless they're complete sentences. It looks like most of our docs take this approach too (although we're not totally consistent).

File: docs/reference/search/knn-search.asciidoc (hunk 7 above)
Comment (jtibshirani) [reply docs/reference/search/knn-search.asciidoc:99]: I think it's clear it needs to be a dense vector value, since those are just arrays of floats?

Taking @mayya-sharipova's suggestion, I've added a note that it must have the same number of dimensions as the vector fields. I think it's good practice to document all the requirements, even if they're pretty self-evident.

File: docs/reference/search/knn-search.asciidoc (hunk 16 above)
Comment (jtibshirani) [reply docs/reference/search/knn-search.asciidoc:114]: I can reorder these to be consistent. I was trying to emphasize that `fields` is what you usually want, so people don't accidentally start using `docvalue_fields`. But there's probably better ways to do this, like through examples and "how to" guides.

File: docs/reference/search/knn-search.asciidoc (hunk 12 above)
Comment (jtibshirani) [reply docs/reference/search/knn-search.asciidoc:124]: This is a subtle point where I think "is determined by" is more accurate/ clear -- the `_score` calculation is not that meaningful. It's a transformed version of the similarity that makes sure to obey Elasticsearch's score requirements.

File: docs/reference/search/knn-search.asciidoc (hunk 16 above)
Comment (jtibshirani) [reply docs/reference/search/knn-search.asciidoc:114]: Oh I just realized that if they were truly alphabetical, then `fields` and `docvalue_fields` would float above `knn`! I think that would really be unfortunate, as the `knn` section contains information that's important to understanding the API.

File: docs/reference/search/knn-search.asciidoc [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+  "knn": {
+    "field": "image_vector",
+    "query_vector": [0.3, 0.1, 1.2],
+    "k": 10,
+    "num_candidates": 100
+  },
+  "_source": ["name", "date"]
+}
+----
+// TEST[setup:my_index]
```
[the remark this answers is not in the document]
Comment (jtibshirani) [reply docs/reference/search/knn-search.asciidoc:38]: Thanks for catching these issues! I ended up hiding the mapping set-up.

> Is there a reason we went with _source instead of fields? Not a huge deal, but I figured fields would be preferred.

I initially had `fields` in some API examples, and one of the readers got very confused. Approaching the API for the first time, they thought `fields` referred to the vector fields for kNN. Unfortunately `fields` is not a widely-used parameter yet, so users would be learning two different concepts at the same time.

File: docs/reference/search/knn-search.asciidoc (hunk 9 above)
Comment (jtibshirani) [reply docs/reference/search/knn-search.asciidoc:108]: This is a nice idea I added the sentence " {es} collects `num_candidates` results from each shard, then merges
them to find the top `k` results". In the high level guide, I think we can step through how the kNN search is executed in more detail.

**Comment by jtibshirani** [comment]
@jrodewig @mayya-sharipova thanks for the great comments. I tried to either respond to them all or address them in the latest commits. I plan to merge later today unless you have more feedback.

File: docs/reference/search/knn-search.asciidoc [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+
+[[knn-search-api-path-params]]
+==== {api-path-parms-title}
+
+`<target>`::
+(Optional, string) Comma-separated list of data streams, indices, and aliases
+to search. Supports wildcards (`*`). To search all data streams and indices,
+use `*` or `_all`.
+
+NOTE: The kNN search API does not support {ccs}.
```
Comment (jimczi) [inline docs/reference/search/knn-search.asciidoc:84]: Is it something that we forbid explicitly ? I didn't test but it should work transparently since we use the search action internally.

Comment (jtibshirani) [reply docs/reference/search/knn-search.asciidoc:84]: It indeed should already work transparently. I wrote this because I didn't want to set a precedent that it supports CCS while we're still figuring out the execution strategy -- especially around how to combine kNN results with term-based results. However it's not something we forbid explicitly (given our strategy of translating at the REST layer, it's not very simple to do). What are your thoughts?

Comment (jimczi) [reply docs/reference/search/knn-search.asciidoc:84]: IMO we're protected by the experimental status. CCS works out of the box and there's nothing that prevents us from removing the support later on. I don't see why we wouldn't though so I am not sure we need the NOTE at the moment.

Comment (jtibshirani) [reply docs/reference/search/knn-search.asciidoc:84]: That reasoning makes sense to me, I'll remove the note to keep things simple.
